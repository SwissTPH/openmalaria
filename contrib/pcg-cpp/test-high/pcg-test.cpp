/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
