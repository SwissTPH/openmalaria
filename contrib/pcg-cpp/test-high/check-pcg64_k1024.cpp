#define RNG pcg64_k1024
#define TWO_ARG_INIT 1
#define AWKWARD_128BIT_CODE 1

#include "pcg-test.cpp"

