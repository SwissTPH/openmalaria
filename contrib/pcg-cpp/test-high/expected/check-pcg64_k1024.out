pcg64_k1024:
      -  result:      64-bit unsigned int
      -  period:      2^65664   (* 2^127 streams)
      -  size:        8224 bytes

Round 1:
  64bit: 0x8d9706afa56d17d4 0x44a818481d01d2df 0x3a13d5e1754b6a96
	 0xc21dcf2b4a6a4655 0x59c706aa47a07c78 0x7f0fc2a85642e319
  Again: 0x8d9706afa56d17d4 0x44a818481d01d2df 0x3a13d5e1754b6a96
	 0xc21dcf2b4a6a4655 0x59c706aa47a07c78 0x7f0fc2a85642e319
  Coins: HHTHHHTHHTHHTHHHHTHHTTTHTTHTHHTTHTHHTTHTTTHHHHHHHTHTTHHHTTHTHTTHH
  Rolls: 4 5 4 3 2 4 3 5 1 3 6 4 3 6 6 1 6 4 6 4 6 1 5 3 6 1 5 2 4 6 1 3 6
   -->   rolling dice used 33 random numbers
  Cards: 3h 5d 8s Kh 8d 9c 7s 2h 6c 6s Ac 8h Jc Ks Qh Td 5h 4d 4h 9h Ah Th
	 6d Jd 2d 3s 8c 2c 3d Js 7d 7c 4s 3c Ts 9s 7h 6h 5s 9d 4c Kc Qs 2s
	 Kd 5c Jh Qc Ad As Qd Tc

Round 2:
  64bit: 0x50638899fce02849 0xb7ce1bddfaba4e10 0x64b3829fcd739931
	 0x66fd9a2569fcd08a 0x96976fd1c0d8cbb4 0xcdf58a9c465c4abf
  Again: 0x50638899fce02849 0xb7ce1bddfaba4e10 0x64b3829fcd739931
	 0x66fd9a2569fcd08a 0x96976fd1c0d8cbb4 0xcdf58a9c465c4abf
  Coins: THTTHHTTHHTHHHTHTHHTTHTHHTTHTHHHHHHHHTTHTTHHTTTHHHTTHHHHTTHHTTTTT
  Rolls: 5 1 2 1 6 5 4 5 4 5 4 2 3 1 3 1 5 1 5 3 2 1 2 3 6 5 5 3 1 2 5 4 3
   -->   rolling dice used 33 random numbers
  Cards: Td Ad 2c 7c 7h Jc 3s 6s Ts 8s 9h 4s Qd Ah 4h Ks Tc 6d 6c Qc 2d Jd
	 Kd 2s 6h 7s 3d 5h Ac 4c Kc 9c 2h 9d 5s 8h Qs 7d 8d Jh 3h 5c 4d Qh
	 3c As 8c 9s Kh Js Th 5d

Round 3:
  64bit: 0xeaceeac7d52650bd 0xf38d15184cc47e50 0xbf1ad446c85f7e68
	 0x4dbcfd5bc51f94ef 0x78d33a2bb1990870 0x2db896fdc9338bdb
  Again: 0xeaceeac7d52650bd 0xf38d15184cc47e50 0xbf1ad446c85f7e68
	 0x4dbcfd5bc51f94ef 0x78d33a2bb1990870 0x2db896fdc9338bdb
  Coins: HTTTTHHTHTHHHHTHTTTHTHTHHHHTHHHTTHTTTHTHTTTTTTHTHTHHTTHTTHHHHTTTH
  Rolls: 5 2 2 1 6 5 5 5 6 3 4 3 1 3 5 2 2 5 5 2 3 6 3 5 6 5 4 2 5 5 2 6 3
   -->   rolling dice used 33 random numbers
  Cards: Tc Th 6d 8s Qd 4s 9h 9d Kc 3c 8c 8h 5h Ah Ks Ts 2d Kh 4c Qh 7s 9c
	 4h 2s Jd 7h 9s 3h 6h Td As 5s Ac 2c 6c Ad Jh Js 6s 5c 3d 4d 2h Kd
	 7d Jc Qs 3s 5d Qc 7c 8d

Round 4:
  64bit: 0xb0a33faf71a1f334 0x94f66590597ae5d8 0x3c7fe68cd26ebdf8
	 0xa2180af6a8d1473c 0xc2688c12707e3b62 0xda672a42428bf0a4
  Again: 0xb0a33faf71a1f334 0x94f66590597ae5d8 0x3c7fe68cd26ebdf8
	 0xa2180af6a8d1473c 0xc2688c12707e3b62 0xda672a42428bf0a4
  Coins: THHTHTHHTTHTTTTTHTTHTTTHTTTHHHHTHTTTTHTTTTTTHHTTTHHTHHTHTHHHTTHHT
  Rolls: 2 6 4 1 2 5 3 2 2 4 1 5 3 3 1 3 6 3 1 1 2 1 5 4 3 1 2 2 2 2 5 2 2
   -->   rolling dice used 33 random numbers
  Cards: 4h 8s 4c Ad 9d Qh 2s Ts Jd Jh 6s 7h 9s Qc 2c As 8h 5d Ac 3d 8c Kc
	 8d 5c Js 2d 9c Qs 3s 3c Td Kh 7s 7c 6d Ah 5h Jc Th 7d 4d Ks 9h 4s
	 5s 2h 6c Kd 6h 3h Tc Qd

Round 5:
  64bit: 0x0e46cf5648162dac 0x085cfe3fc86e06af 0x7aa43a225abaf9c1
	 0x898e076f83632286 0xaf1f07f02715ef10 0x9807752b8f8907b6
  Again: 0x0e46cf5648162dac 0x085cfe3fc86e06af 0x7aa43a225abaf9c1
	 0x898e076f83632286 0xaf1f07f02715ef10 0x9807752b8f8907b6
  Coins: HHTTHHHTTTHTTTHTTTHTTHHTTHTHTHHHTHTHHTHTHHTHHHHTHTHHHHTHTTTTHTHHT
  Rolls: 1 4 1 1 4 1 6 3 1 2 6 4 3 4 2 1 1 5 4 4 2 3 6 1 5 2 4 6 5 3 2 1 3
   -->   rolling dice used 33 random numbers
  Cards: Jc Ad Ts Qs Jh 7d 2h Ah 2d 4s Kc Qc 6s 5h 7s 6h Qd 3d 4d Jd 4c Js
	 6c 4h 9d 8c 2s As 8s Ks 8h Kh 3h 3c 8d Th 9h Tc 7h 5d 6d 9s Td 9c
	 5c Ac 7c 5s 2c Kd Qh 3s

